go/teams/hidden_loader_test.go

```diff
@@ -0,0 +1,1113 @@
+			if hiddenResp != nil {
+				hiddenResp.RespType = libkb.MerkleHiddenResponseTypeABSENCEPROOF
+			}
+			return r1, r2, hiddenResp, err
+		},
+	}
+	tcs[1].G.SetFastTeamLoader(ftl)
+
+	res, err = tcs[1].G.GetFastTeamLoader().Load(libkb.NewMetaContextForTest(*tcs[1]), keybase1.FastTeamLoadArg{
+		ID:                   teamID,
+		ForceRefresh:         true,
+		Applications:         []keybase1.TeamApplication{keybase1.TeamApplication_CHAT},
+		KeyGenerationsNeeded: []keybase1.PerTeamKeyGeneration{keybase1.PerTeamKeyGeneration(2)},
+	})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "Server claimed (and proved) there are no committed hidden chain links in the chain, but we had previously seen a committed link with seqno 1")
+
+	// now load the team again, but this time we change the response of the server and change the proof type
+	ftl.world = CorruptingMockLoaderContext{
+		LoaderContext: world,
+		merkleCorruptorFunc: func(r1 keybase1.Seqno, r2 keybase1.LinkID, hiddenResp *libkb.MerkleHiddenResponse, err error) (keybase1.Seqno, keybase1.LinkID, *libkb.MerkleHiddenResponse, error) {
+			if hiddenResp != nil {
+				hiddenResp.RespType = libkb.MerkleHiddenResponseTypeNONE
+			}
+			return r1, r2, hiddenResp, err
+		},
+	}
+	tcs[1].G.SetFastTeamLoader(ftl)
+
+	res, err = tcs[1].G.GetFastTeamLoader().Load(libkb.NewMetaContextForTest(*tcs[1]), keybase1.FastTeamLoadArg{
+		ID:                   teamID,
+		ForceRefresh:         true,
+		Applications:         []keybase1.TeamApplication{keybase1.TeamApplication_CHAT},
+		KeyGenerationsNeeded: []keybase1.PerTeamKeyGeneration{keybase1.PerTeamKeyGeneration(2)},
+	})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "hidden chain data missing error: the server did not return the necessary hidden chain data")
+}
+
+func TestFTLFailsIfHiddenTailIsTamperedBeforeFirstLoad(t *testing.T) {
+	fus, tcs, cleanup := setupNTests(t, 4)
+	defer cleanup()
+
+	t.Logf("create team")
+	teamName, teamID := createTeam2(*tcs[0])
+
+	t.Logf("add B and C to the team so they can load it")
+	_, err := AddMember(context.TODO(), tcs[0].G, teamName.String(), fus[1].Username, keybase1.TeamRole_WRITER, nil)
+	require.NoError(t, err)
+	_, err = AddMember(context.TODO(), tcs[0].G, teamName.String(), fus[3].Username, keybase1.TeamRole_WRITER, nil)
+	require.NoError(t, err)
+
+	res, err := tcs[1].G.GetFastTeamLoader().Load(libkb.NewMetaContextForTest(*tcs[1]), keybase1.FastTeamLoadArg{
+		ID:                   teamID,
+		ForceRefresh:         true,
+		Applications:         []keybase1.TeamApplication{keybase1.TeamApplication_CHAT},
+		KeyGenerationsNeeded: []keybase1.PerTeamKeyGeneration{keybase1.PerTeamKeyGeneration(1)},
+	})
+	require.NoError(t, err)
+	require.Equal(t, res.Name.String(), teamName.String())
+	require.Equal(t, 1, len(res.ApplicationKeys))
+	require.EqualValues(t, 1, res.ApplicationKeys[0].KeyGeneration)
```

Comment:
is it possible to factor out any of your test code into reusable blocks? I find that often helps over the long run.